From cc1d56c1f4c63645f9682db6634c77e72b810f47 Mon Sep 17 00:00:00 2001
From: Ky-Anh Huynh <kyanh@theslinux.org>
Date: Sun, 18 Aug 2013 22:46:54 +0700
Subject: [PATCH 3002/3002] Remove :state icon on systray

The systray has two groups (aka :branches) of icons on systray.
One is to list all IMs, the other is to indicate the state of the
current IM (on, off, or direct_input). There may be some other
group but I do not know.

Because we tell :uim to switch to :direct_input as an alternative
of :toggle action, we do not really need a second icon to show
the same thing. We just remove them, and this will make the icon
bigger/viewable on the systray. The combination "^ Alt \" is
still valid: it can be used to switch :telex <-> :vni. Note that
some user option may make new feature buggy. Also note that we
only test this feature on :Gtk2, but :Gtk3 should work, too.

I spent my two days (sat & sun) to figure out how I can patch.
I first thought there are some Gtk code that creates two groups.
Unfortunately that is completely wrong. The :toolbar reads the
instructions from :uim_fd (file descriptor that contains control
messages), and parses the messages to create buttons and icons
for the systray. See (helper_toolbar_prop_list_update). That
is why you see lines that use (strcmp) to check if the current
message/line wants to create a :branch (:group) or :leaf (icon)
---
 uim-vi/src/uim-1.8.5/gtk2/toolbar/common-gtk.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/uim-vi/src/uim-1.8.5/gtk2/toolbar/common-gtk.c b/uim-vi/src/uim-1.8.5/gtk2/toolbar/common-gtk.c
index 2d2a97f..30bdf7a 100644
--- a/uim-vi/src/uim-1.8.5/gtk2/toolbar/common-gtk.c
+++ b/uim-vi/src/uim-1.8.5/gtk2/toolbar/common-gtk.c
@@ -781,6 +781,10 @@ helper_toolbar_prop_list_update(GtkWidget *widget, gchar **lines)
 	indication_id = cols[1];
 	iconic_label  = safe_gettext(cols[2]);
 	tooltip_str   = safe_gettext(cols[3]);
+        if (strcmp("direct_input", indication_id)
+            && strcmp("on", indication_id)
+            && strcmp("off", indication_id))
+        {
           button = prop_button_create(widget,
                     indication_id, iconic_label, tooltip_str);
           append_prop_button(widget, button);
@@ -789,6 +793,7 @@ helper_toolbar_prop_list_update(GtkWidget *widget, gchar **lines)
               || g_str_has_suffix(indication_id, "_direct"))) {
             is_hidden = TRUE;
           }
+        }
       } else if (!strcmp("leaf", cols[0]) && has_n_strs(cols, 7)) {
 	indication_id = cols[1];
 	iconic_label  = safe_gettext(cols[2]);
@@ -796,10 +801,15 @@ helper_toolbar_prop_list_update(GtkWidget *widget, gchar **lines)
 	tooltip_str   = safe_gettext(cols[4]);
 	action_id     = cols[5];
 	is_selected   = cols[6];
+        if (strcmp("direct_input", indication_id)
+            && strcmp("on", indication_id)
+            && strcmp("off", indication_id))
+        {
           prop_button_append_menu(button,
                 indication_id, label, tooltip_str, action_id,
                 is_selected);
         }
+      }
       g_strfreev(cols);
     }
   }
-- 
1.8.3

